project-koku/koku
# skip polling providers still processing

## Jira Ticket

[COST-5180](https://issues.redhat.com/browse/COST-5180)

## Description

This change will add an additional filter to polling batch collection to not add provider to batch that has not completed yet (XL customers that take longer than 24 hours to process)

## Testing

1. Checkout Branch
2. Restart Koku
3. Make sure you set your POLLING_TIMER appropriately (E.I NOT 1 second)
4. Load AWS data (everything should process correct)
5. Hit [download](http://localhost:5042/api/cost-management/v1/download/) - Nothing new should attempt to be queued
6. Run the following SQL (Change uuid for your provider)
7. Run # update api_provider set polling_timestamp = '2024-06-23 00:00:00+00' where uuid = '9dfc1842-d654-42d8-8e70-efc49b33fa5f';
8. Hit download again and observe provider is polled
9.  Run # update api_provider set data_updated_timestamp = '2024-06-17 00:00:00+00' where uuid = '9dfc1842-d654-42d8-8e70-efc49b33fa5f';
10. Hit download and observe provider not added to polling (Since data_updated_timestamp is older than polling time and the polling time is within 24 hours, this simulates a large provider still processing)
11. Run # update api_provider set polling_timestamp = '2024-06-23 00:00:00+00' where uuid = '9dfc1842-d654-42d8-8e70-efc49b33fa5f';
12. Hit download, (This time we should collect the provider since the updated timestamp is outside our 7 day window and the polling time is outside 24 hours)
13. Run # update api_provider set polling_timestamp = null where uuid = '9dfc1842-d654-42d8-8e70-efc49b33fa5f';
14. Download again and see polling for provider (This simulates a new provider that has no polling time set yet)

## Release Notes

```markdown
* [COST-5180](https://issues.redhat.com/browse/COST-5180) Only attempt to fetch new reports if current processing is complete
```


Notes:

I left a TODO in here around checking data_updated_timestamp being older and not getting updated. For example if for any reason our download processing fails, maybe we kill tasks or something. It's conceivable that the data_updated_timestamp wont be updated, meaning we never poll that provider again without manual intervention.
CORRECTION I added the following to handle this:

`| Q(data_updated_timestamp__lte=process_wait_delta)`

I have some interesting metrics on this if people are interested. We currently have 62 providers in that state 28 of those are currently processing though! The rest (34) we just keep queueing up but have not completed in years. Probably in bad states.
